src/raft.c

```diff
@@ -97,24 +100,211 @@ void entryAttachRaftReq(RedisRaftCtx *rr, raft_entry_t *entry, RaftReq *req)
+            if (RedisModule_KeyExists(rr->ctx, cmd->argv[keyindex[j]])) {
+                found++;
+
+                /* test if locked */
+                int nokey;
+                RedisModule_DictGet(rr->locked_keys, cmd->argv[keyindex[j]], &nokey);
+                if (!nokey) {
+                    RedisModule_Free(keyindex);
+                    return SomeExist;
+                }
+            }
+        }
+        RedisModule_Free(keyindex);
+    }
+
+    if (found != total_keys) {
+        return (found == 0) ? NoneExist : SomeExist;
+    }
+
+    return AllExist;
+}
+
+/* figure out the "owner shardgroup"
+ *
+ * the owner shardgroup is either
+ *
+ * 1) a shardgroup that owns the slot as a stable slot
+ * 2) a shardgroup that owns the slot as a migrating slot.  by definition is this shardgroup is local, then it
+ *    can't also be importing (a single RedisRaft cluster cannot be both importing and migrating the same slot
+ * 3) a shardgroup marked as local (i.e. corresponding to this cluster) that owns the slot as an importing slot and
+ *    has a RaftRedisCommandArray marked as asking
+ */
+
+static ShardGroup *GetSlotShardGroup(RedisRaftCtx *rr, unsigned int slot, bool asking)
+{
+    ShardGroup *sg = rr->sharding_info->stable_slots_map[slot];
+    if (sg != NULL) {
+        return sg;
+    }
+
+    sg = rr->sharding_info->migrating_slots_map[slot];
+    if (sg && sg->local) {
+        return sg;
+    }
+
+    if (asking) {
+        ShardGroup *isg = rr->sharding_info->importing_slots_map[slot];
+        if (isg && isg->local) {
+            return isg;
+        }
+    }
+
+    return sg;
+}
+
+static RRStatus validateRaftRedisCommandArray(RedisRaftCtx *rr, RedisModuleCtx *reply_ctx,
+                                              RaftRedisCommandArray *cmds, unsigned int slot)
+{
+    RedisModule_Assert(slot <= REDIS_RAFT_HASH_MAX_SLOT);
+
+    /* Make sure hash slot is mapped and handled locally. */
+    ShardGroup *sg = GetSlotShardGroup(rr, slot, cmds->asking);
+    if (!sg) {
+        if (reply_ctx) {
+            RedisModule_ReplyWithError(reply_ctx, "CLUSTERDOWN Hash slot is not served");
+        }
+        return RR_ERROR;
+    }
+
+    SlotRangeType slot_type = SLOTRANGE_TYPE_UNDEF;
+    for (size_t i = 0; i < sg->slot_ranges_num; i++) {
+        if (sg->slot_ranges[i].start_slot <= slot && slot <= sg->slot_ranges[i].end_slot) {
+            slot_type = sg->slot_ranges[i].type;
+            break;
+        }
+    }
+
+    if (slot_type == SLOTRANGE_TYPE_UNDEF) {
+        if (reply_ctx) {
+            RedisModule_ReplyWithError(reply_ctx, "ERR internal error, couldn't associate a shardgroup slot to this request");
+        }
+        return RR_ERROR;
+    }
+
+    if (!sg->local) {
+        if (reply_ctx) {
+            sg->next_redir++;
+            if (sg->next_redir >= sg->nodes_num) {
+                sg->next_redir = 0;
+            }
+            replyRedirect(reply_ctx, slot, &sg->nodes[sg->next_redir].addr);
```

Comment:
```suggestion
            replyRedirect(reply_ctx, slot, &sg->nodes[sg->next_redir].addr);
            sg->next_redir = (sg->next_redir + 1) % sg->nodes_num;
```
We can use modulus here right?. This change starts redirecting from node 0, just beware. If not desired, you can move modulus up